File: server/src/main/java/gg/mineral/server/network/packet/login/clientbound/LoginDisconnectPacket.java
package gg.mineral.server.network.packet.login.clientbound;

import dev.zerite.craftlib.chat.component.BaseChatComponent;
import dev.zerite.craftlib.chat.component.ChatComponentUtil;
import gg.mineral.api.network.packet.Packet;

import io.netty.buffer.ByteBuf;

public final record LoginDisconnectPacket(BaseChatComponent baseChatComponent) implements Packet.OUTGOING {
    @Override
    public void serialize(ByteBuf os) {
        writeString(os, ChatComponentUtil.toJson(baseChatComponent));
    }

    @Override
    public byte getId() {
        return 0x00;
    }
}

File: server/src/main/kotlin/gg/mineral/server/network/packet/login/clientbound/LoginDisconnectPacket.kt
package gg.mineral.server.network.packet.login.clientbound

import gg.mineral.api.network.packet.Packet
import io.netty.buffer.ByteBuf
import net.md_5.bungee.api.chat.BaseComponent

@JvmRecord
data class LoginDisconnectPacket(val components: Array<out BaseComponent>) : Packet.Outgoing {
    override fun serialize(os: ByteBuf) =
        os.writeString(VersionedComponentSerializer.getDefault().toString(components))

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as LoginDisconnectPacket

        return components.contentEquals(other.components)
    }

    override fun hashCode(): Int {
        return components.contentHashCode()
    }

    override val id: Byte
        get() = 0x00
}
